src/algebra/group_with_zero.lean

```diff
@@ -0,0 +1,504 @@
+/-
+Copyright (c) 2020 Johan Commelin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johan Commelin
+-/
+
+import algebra.group.units
+
+/-!
+# Groups with an adjoined zero element
+
+This file describes structures that are not usually studied on their own right in mathematics,
+namely a special sort of monoid: apart from a distinguished “zero element” they form a group,
+or in other words, they are groups with an adjoined zero element.
+
+Examples are:
+
+* division rings;
+* the value monoid of a multiplicative valuation;
+* in particular, the non-negative real numbers.
+
+## Main definitions
+
+* `group_with_zero`
+* `comm_group_with_zero`
+
+## Implementation details
+
+As is usual in mathlib, we extend the inverse function to the zero element,
+and require `0⁻¹ = 0`.
+
+-/
+
+set_option old_structure_cmd true
+
+section prio
+set_option default_priority 10 -- see Note [default priority]
+
+/-- A type `M` is a “monoid with zero” if it is a monoid with zero element. -/
+class monoid_with_zero (G : Type*) extends monoid G, mul_zero_class G.
+
+/-- A type `M` is a commutative “monoid with zero”
+if it is a commutative monoid with zero element. -/
+class comm_monoid_with_zero (G : Type*) extends monoid G, mul_zero_class G.
+
+/-- A type `G` is a “group with zero” if it is a monoid with zero element (distinct from `1`)
+such that every nonzero element is invertible.
+The type is required to come with an “inverse” function, and the inverse of `0` must be `0`.
+
+Examples include division rings and the ordered monoids that are the
+target of valuations in general valuation theory.-/
+class group_with_zero (G : Type*)
+  extends monoid_with_zero G, has_inv G, zero_ne_one_class G :=
```

Comment:
I think I would have been tempted not to use `G` for the notation, and would have gone for $G_0$ or $G^0$ or whatever. Or even `\alpha` ;-)

Reply:
Fixed. It's now G_0 everywhere.